File: src/searching/cpp/binary_search/src/binary_search.cpp
#include "binary_search.h"

#include <cstddef>
#include <optional>
#include <vector>

template <typename T>
std::optional<std::size_t> binary_search(const std::vector<T> &arr, const T &target) {
  if (arr.empty()) {
    return std::nullopt;
  }

  std::size_t low = 0;
  std::size_t high = arr.size() - 1;

  while (low <= high) {
    std::size_t mid = low + (high - low) / 2;
    if (arr[mid] == target) {
      return mid;
    } else if (arr[mid] < target) {
      low = mid + 1;
    } else {
      if (mid == 0) {
        break;
      }
      high = mid - 1;
    }
  }

  return std::nullopt;
}

// Explicit template instantiations
template std::optional<std::size_t> binary_search<int>(const std::vector<int> &arr, const int &target);
template std::optional<std::size_t> binary_search<float>(const std::vector<float> &arr, const float &target);
template std::optional<std::size_t> binary_search<double>(const std::vector<double> &arr, const double &target);
template std::optional<std::size_t> binary_search<char>(const std::vector<char> &arr, const char &target);

File: src/searching/python/binary_search/src/binary_search.py
"""Binary search algorithm implementation."""

from typing import List, Optional, TypeVar

T = TypeVar("T")


def binary_search(array: List[T], target: T) -> Optional[int]:
    """
    Search for a target value in a sorted array using binary search.

    This algorithm repeatedly halves the search interval by comparing the
    target to the middle element until found or the interval is empty.

    Time complexity: O(log n)
    Space complexity: O(1)

    Args:
        array: The sorted list of comparable elements to search.
        target: The value to search for.

    Returns:
        The index of the target if found, None otherwise.

    Note:
        The array must be sorted in ascending order for correct results.
    """
    low = 0
    high = len(array) - 1

    while low <= high:
        mid = (low + high) // 2
        if array[mid] == target:
            return mid
        elif array[mid] < target:
            low = mid + 1
        else:
            high = mid - 1

    return None
